### ISSUE TITLE
copy of current selected range to compare with new range in cellselectionmodel

### ISSUE THREAD

**Issue by kth999** [6pac/SlickGrid#341]
In the function handleKeyDown(e) of slick.cellselectionmodel.js,
I think the exising selected range should be copied to be compared properly in setSelectedRanges function.

the line 133
`ranges = getSelectedRanges();`
should be changed like
`ranges = getSelectedRanges().slice(0);`

**Comment by ghiscoding** [comment]
Can you provide more details about the problem? I see you're doing a deep copy but perhaps we should check that we do have an array before calling `slice`, this might throw some errors if the function call returns null.

**Comment by kth999** [comment]
The `ranges` variable returned from `getSelectedRanges()` is the member variable, which is being passed to the `setSelectedRanges` function on line 172. In the `setSelectedRanges`, the same member variable is being compared to itself, resulting `rangeHasChanged` false.
So, I can't change range using keyboard with shiftKey.

**Comment by SatanEnglish** [comment]
@ghiscoding  Think what i'm seeing is related to this.
```
var selectionModel = new Slick.RowSelectionModel({selectActiveRow: true}; <== TRUE is the important thing here.
grid.setSelectionModel(selectionModel);
```
Old version would allow you to select each row by the check box and have multiple selected.
New version does not 

`   
     if (lookup[row] !== _selectedRowsLookup[row]) 
    {
          _grid.invalidateRow(row);
          delete _selectedRowsLookup[row];
        }
`
The above is deleting the row even tho it is the same row.

@kth999
I just found the issue for mine!!
Try commenting out the following lines.
`        function toggleRowSelection(row) {
            if (_selectedRowsLookup[row]) {
                _grid.setSelectedRows($.grep(_grid.getSelectedRows(), function (n) {
                    return n != row
                }));
            } else {
                _grid.setSelectedRows(_grid.getSelectedRows().concat(row));
            }
            //_grid.setActiveCell(row, getCheckboxColumnCellIndex());
            //_grid.focus();
        }`

The set and then focus seems to be acting like a double click on the checkbox

**Comment by ghiscoding** [comment]
@kth999 
Can you provide 1 of the multiple [examples](https://github.com/6pac/SlickGrid/wiki/Examples) to replicate the issue you're having? ...or at least some of the grid options you're using to replicate the issue. Without a way to replicate, we can't really fix the issue (or at least we don't want to change the code without a valid reason). 

So far I can't replicate what you said, if I take this online [example - checkbox row select](http://6pac.github.io/SlickGrid/examples/example-checkbox-row-select.html), you can see (in the animated gif) that it works when I'm holding "Shift" key and the click with the mouse to select the range. So... where's the issue? I can't find it

![smbx7uhlyo](https://user-images.githubusercontent.com/643976/53615101-6070e000-3baa-11e9-9175-0874b75b8a5f.gif)

**Comment by SatanEnglish** [comment]
@ghiscoding 
My bad I've confused you he's using `slick.cellselectionmodel.js` which is not the selection model used in the example you tested.

e.g.  

```diff
+ grid2.setSelectionModel(new Slick.CellSelectionModel());
- grid2.setSelectionModel(new Slick.RowSelectionModel({selectActiveRow: false}));
```

He also says hes using the shift and DOWN key you example is clicking.

I had a quick play and holding shift and pressing down doesn't do anything. (Though not shore what it should do!)
![2019-03-01_20-32-38](https://user-images.githubusercontent.com/2100563/53623160-2adafe00-3c61-11e9-9c39-957825dada82.gif)


With his line it does what it does in the select model in example. (Kind of)
![2019-03-01_20-31-46](https://user-images.githubusercontent.com/2100563/53623128-0da62f80-3c61-11e9-9ab2-b198d9a830b6.gif)

I guess it should do something but not sure what. (Check if the row is the checkbox row and tick if so...)

**Comment by ghiscoding** [comment]
Hmm ok I misunderstood then, it wasn't clear, I don't think it was mentioned, that it was SHIFT + Arrows Keys that were used. I'll have another look later then

**Comment by ghiscoding** [comment]
@kth999 
Merged PR #348 that addresses the issue you brought up, thanks for your feedback and troubleshooting.
